emod-abm JSONDecodeError

Issue with [Merge pull request #109 from UM-Bridge/issue-arm-builds](https://github.com/UM-Bridge/benchmarks/actions/runs/22483113582/job/65126097041)

See error here:

```
Run docker run --network=host -e model_host=http://localhost:4242 linusseelinger/testing-protocol-conformity-current:latest
Unable to find image 'linusseelinger/testing-protocol-conformity-current:latest' locally
latest: Pulling from linusseelinger/testing-protocol-conformity-current
2ca1bfae7ba8: Pulling fs layer
b4bef873be5f: Pulling fs layer
1ef3105d0b8c: Pulling fs layer
1ef3105d0b8c: Verifying Checksum
1ef3105d0b8c: Download complete
b4bef873be5f: Verifying Checksum
b4bef873be5f: Download complete
2ca1bfae7ba8: Verifying Checksum
2ca1bfae7ba8: Download complete
2ca1bfae7ba8: Pull complete
b4bef873be5f: Pull complete
1ef3105d0b8c: Pull complete
Digest: sha256:b29252f7977fa2806396508c9c2a9e7acf4bf49f89b35fbb8cc29273a41f4c8c
Status: Downloaded newer image for linusseelinger/testing-protocol-conformity-current:latest
============================= test session starts ==============================
platform linux -- Python 3.13.5, pytest-8.3.5, pluggy-1.5.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /
collecting ... collected 22 items

test_model_schema.py::test_connection[http://localhost:4242] PASSED      [  4%]
test_model_schema.py::test_evaluate[http://localhost:4242-0.0] FAILED    [  9%]
test_model_schema.py::test_evaluate_with_wrong_model_name[http://localhost:4242] PASSED [ 13%]
test_model_schema.py::test_evaluate_unsupported[http://localhost:4242] PASSED [ 18%]
test_model_schema.py::test_evaluate_with_wrong_input_dimensions[http://localhost:4242-0.0] PASSED [ 22%]
test_model_schema.py::test_gradient[http://localhost:4242-0.0] PASSED    [ 27%]
test_model_schema.py::test_gradient_unsupported[http://localhost:4242] PASSED [ 31%]
test_model_schema.py::test_gradient_with_wrong_input_dimensions[http://localhost:4242-0.0] PASSED [ 36%]
test_model_schema.py::test_gradient_with_wrong_inwrt[http://localhost:4242-0.0] PASSED [ 40%]
test_model_schema.py::test_gradient_with_wrong_outwrt[http://localhost:4242-0.0] PASSED [ 45%]
test_model_schema.py::test_apply_jacobian[http://localhost:4242-0.0] PASSED [ 50%]
test_model_schema.py::test_apply_jacobian_unsupported[http://localhost:4242] PASSED [ 54%]
test_model_schema.py::test_apply_jacobian_with_wrong_input_dimensions[http://localhost:4242-0.0] PASSED [ 59%]
test_model_schema.py::test_apply_hessian[http://localhost:4242-0.0] PASSED [ 63%]
test_model_schema.py::test_apply_hessian_unsupported[http://localhost:4242] PASSED [ 68%]
test_model_schema.py::test_apply_hessian_with_wrong_input_dimensions[http://localhost:4242-0.0] PASSED [ 72%]
test_model_schema.py::test_input_sizes[http://localhost:4242] PASSED     [ 77%]
test_model_schema.py::test_output_sizes[http://localhost:4242] PASSED    [ 81%]
test_model_schema.py::test_protocol_version[http://localhost:4242] PASSED [ 86%]
test_model_schema.py::test_info[http://localhost:4242] PASSED            [ 90%]
test_model_schema.py::test_model_info[http://localhost:4242] PASSED      [ 95%]
test_model_schema.py::test_model_info_with_wrong_model_name[http://localhost:4242] PASSED [100%]

=================================== FAILURES ===================================
___________________ test_evaluate[http://localhost:4242-0.0] ___________________

self = <Response [200]>, kwargs = {}

    def json(self, **kwargs):
        r"""Returns the json-encoded content of a response, if any.
    
        :param \*\*kwargs: Optional arguments that ``json.loads`` takes.
        :raises requests.exceptions.JSONDecodeError: If the response body does not
            contain valid json.
        """
    
        if not self.encoding and self.content and len(self.content) > 3:
            # No encoding set. JSON RFC 4627 section 3 states we should expect
            # UTF-8, -16 or -32. Detect which one to use; If the detection or
            # decoding fails, fall back to `self.text` (using charset_normalizer to make
            # a best guess).
            encoding = guess_json_utf(self.content)
            if encoding is not None:
                try:
                    return complexjson.loads(self.content.decode(encoding), **kwargs)
                except UnicodeDecodeError:
                    # Wrong UTF codec detected; usually because it's not UTF-8
                    # but some other 8-bit codec.  This is an RFC violation,
                    # and the server didn't bother to tell us what codec *was*
                    # used.
                    pass
                except JSONDecodeError as e:
                    raise RequestsJSONDecodeError(e.msg, e.doc, e.pos)
    
        try:
>           return complexjson.loads(self.text, **kwargs)

usr/lib/python3/dist-packages/requests/models.py:974: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
usr/lib/python3.13/json/__init__.py:346: in loads
    return _default_decoder.decode(s)
usr/lib/python3.13/json/decoder.py:345: in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <json.decoder.JSONDecoder object at 0x7f7382c13cb0>
s = '{"output": [[{\'New Infections\': {\'Data\': [0]}, \'Infected\': {\'Data\': [0]}, \'Infectious Population\': {\'Data\...Population\': {\'Data\': [0]}, \'Recovered Population\': {\'Data\': [0]}, \'Exposed Population\': {\'Data\': [0]}}]] }'
idx = 0

    def raw_decode(self, s, idx=0):
        """Decode a JSON document from ``s`` (a ``str`` beginning with
        a JSON document) and return a 2-tuple of the Python
        representation and the index in ``s`` where the document ended.
    
        This can be used to decode a JSON document from a string that may
        have extraneous data at the end.
    
        """
        try:
>           obj, end = self.scan_once(s, idx)
E           json.decoder.JSONDecodeError: Expecting property name enclosed in double quotes: line 1 column 15 (char 14)

usr/lib/python3.13/json/decoder.py:361: JSONDecodeError

During handling of the above exception, another exception occurred:

model_url = 'http://localhost:4242', input_value = 0.0

    def test_evaluate(model_url, input_value):
    
        resp_info = requests.get(f'{model_url}/Info')
        assert resp_info.status_code == 200
        model_name = resp_info.json()["models"][0]
    
        input_model_name = {"name": model_name}
        resp_info = requests.post(f'{model_url}/ModelInfo', json=input_model_name)
        if not resp_info.json()["support"]["Evaluate"]:
          return
    
        inputSizesJSON = requests.post(f'{model_url}/InputSizes', json=input_model_name).json()
    
        inputParams = {"input": [], "name": model_name, "config": {}}
        for i in range(0, len(inputSizesJSON["inputSizes"])):
          inputSizesJSON_i = inputSizesJSON["inputSizes"][i]
          inputSizesJSON_len = len(inputSizesJSON["inputSizes"])
          # Handles the case for one input vector
          if hasattr(input_value, '__len__') and inputSizesJSON_len == 1:
            assert len(input_value) == inputSizesJSON_i
            inputParams["input"].append(input_value)
          # The case where there are multiple input vector
          elif hasattr(input_value, '__len__') and inputSizesJSON_len != 1:
            assert len(input_value[i]) == inputSizesJSON_i
            inputParams["input"].append(input_value[i])
          # Single number entered, will be expanded if sizes unmatched
          else:
            inputParams["input"].append([input_value] * inputSizesJSON_i)
            assert len(inputParams["input"][i]) == inputSizesJSON_i
    
        resp = requests.post(f'{model_url}/Evaluate', headers={}, data=json.dumps(inputParams,indent=4))
    
        assert resp.status_code == 200
    
        schema = {
          "type":"object",
          "properties":{
            "output":{
              "type":"array",
              "items": {
                "type":"array",
                "items": {
                  "type": "number",
                },
                "minItems": 1
              },
              "minItems": 1
            }
          },
          "required":[
            "output"
          ],
          "additionalProperties":False
        }
    
>       jsonschema.validate(instance=resp.json(), schema=schema)

test_model_schema.py:73: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Response [200]>, kwargs = {}

    def json(self, **kwargs):
        r"""Returns the json-encoded content of a response, if any.
    
        :param \*\*kwargs: Optional arguments that ``json.loads`` takes.
        :raises requests.exceptions.JSONDecodeError: If the response body does not
            contain valid json.
        """
    
        if not self.encoding and self.content and len(self.content) > 3:
            # No encoding set. JSON RFC 4627 section 3 states we should expect
            # UTF-8, -16 or -32. Detect which one to use; If the detection or
            # decoding fails, fall back to `self.text` (using charset_normalizer to make
            # a best guess).
            encoding = guess_json_utf(self.content)
            if encoding is not None:
                try:
                    return complexjson.loads(self.content.decode(encoding), **kwargs)
                except UnicodeDecodeError:
                    # Wrong UTF codec detected; usually because it's not UTF-8
                    # but some other 8-bit codec.  This is an RFC violation,
                    # and the server didn't bother to tell us what codec *was*
                    # used.
                    pass
                except JSONDecodeError as e:
                    raise RequestsJSONDecodeError(e.msg, e.doc, e.pos)
    
        try:
            return complexjson.loads(self.text, **kwargs)
        except JSONDecodeError as e:
            # Catch JSON-related errors and raise as requests.JSONDecodeError
            # This aliases json.JSONDecodeError and simplejson.JSONDecodeError
>           raise RequestsJSONDecodeError(e.msg, e.doc, e.pos)
E           requests.exceptions.JSONDecodeError: Expecting property name enclosed in double quotes: line 1 column 15 (char 14)

usr/lib/python3/dist-packages/requests/models.py:978: JSONDecodeError
=========================== short test summary info ============================
FAILED test_model_schema.py::test_evaluate[http://localhost:4242-0.0] - reque...
========================= 1 failed, 21 passed in 0.32s =========================
Error: Process completed with exit code 1.
```

--- a/benchmarks/abm-attack-fraction/server.py
+++ b/benchmarks/abm-attack-fraction/server.py
@@ -27,7 +27,8 @@ class EMODBenchmarkModel(umbridge.Model):
         # returns attack fraction of the model ([0,1]), -0 for out of bounds
         model_output = self.model(parameters, config)[0]
 
-        model_output = [1 - model_output[0]['Susceptible Population']['Data'][-1] ]
+        result = json.loads(model_output[0])
+        model_output = [1 - result['Susceptible Population']['Data'][-1]]
 
         print(model_output)
 

--- a/models/emod-abm/model.py
+++ b/models/emod-abm/model.py
@@ -60,7 +60,7 @@ class EMODForwardModel(umbridge.Model):
             channels = ['New Infections', 'Infected', 'Infectious Population', 'Susceptible Population', 'Symptomatic Population', 'Recovered Population', 'Exposed Population']
             values = [0, 0, 0, 1, 0, 0, 0]
             out_of_bounds = {k:{"Data":[v]} for k,v in zip(channels, values)}
-            return [[out_of_bounds]]
+            return [[json.dumps(out_of_bounds)]]
 
         # update config file
         inf_prd_mean   = 8.0
@@ -97,7 +97,8 @@ class EMODForwardModel(umbridge.Model):
         # remove extra files
         self._cleanup()
 
-        return [[{float(c):inset_chart['Channels'][c] for c in ['New Infections', 'Infected', 'Infectious Population', 'Susceptible Population', 'Symptomatic Population', 'Recovered Population', 'Exposed Population']}]]
+        result = {c: inset_chart['Channels'][c] for c in ['New Infections', 'Infected', 'Infectious Population', 'Susceptible Population', 'Symptomatic Population', 'Recovered Population', 'Exposed Population']}
+        return [[json.dumps(result)]]
 
     def supports_evaluate(self):
         return True    

